pkg/geo/wkt/wkt_test.go

```diff
@@ -334,6 +334,219 @@ func TestUnmarshal(t *testing.T) {
+				geom.NewLineStringFlat(geom.XYM, []float64{0, 0, 200, 0.1, -1, -20}),
+				geom.NewPolygonFlat(geom.XYM, []float64{0, 0, 7, 1, -1, -50, 2, 0, 0, 0, 0, 7}, []int{12}),
+				geom.NewMultiPointFlat(geom.XYM, []float64{-1, 5, -16, 0.23, 7, 0}),
+				geom.NewMultiLineStringFlat(geom.XYM, []float64{0, -1, -2, 2, 5, 7}, []int{6}),
+				geom.NewMultiPolygonFlat(geom.XYM, []float64{0, 0, 0, 1, 1, 1, 2, 3, 1, 0, 0, 0}, [][]int{{12}}),
+			),
+		},
+		{
+			desc: "parse 2D+M geometrycollection with nested geometrycollection",
+			equivInputs: []string{
+				"GEOMETRYCOLLECTION(GEOMETRYCOLLECTION M (POINT EMPTY, LINESTRING M (0 0 0, 1 1 1)))",
+				"GEOMETRYCOLLECTION(GEOMETRYCOLLECTION M (POINT M EMPTY, LINESTRING M (0 0 0, 1 1 1)))",
+				"GEOMETRYCOLLECTION(GEOMETRYCOLLECTION(POINT M EMPTY, LINESTRING M (0 0 0, 1 1 1)))",
+				"GEOMETRYCOLLECTION M (GEOMETRYCOLLECTION(POINT M EMPTY, LINESTRING M (0 0 0, 1 1 1)))",
+				"GEOMETRYCOLLECTION M (GEOMETRYCOLLECTION M (POINT M EMPTY, LINESTRING M (0 0 0, 1 1 1)))",
+			},
+			expected: geom.NewGeometryCollection().MustPush(
+				geom.NewGeometryCollection().MustPush(
+					geom.NewPointEmpty(geom.XYM),
+					geom.NewLineStringFlat(geom.XYM, []float64{0, 0, 0, 1, 1, 1}),
+				),
+			),
+		},
+		{
+			desc: "parse 3D geometrycollection",
+			equivInputs: []string{`GEOMETRYCOLLECTION Z (
+POINT Z (2 3 4),
+LINESTRING Z (0 -1 1, 7 -1 -9),
+POLYGON Z ((0 0 7, 1 -1 -50, 2 0 0, 0 0 7)),
+MULTIPOINT Z ((2 3 1), EMPTY),
+MULTILINESTRING Z (EMPTY, EMPTY, (1 1 1, 2 2 2, 3 3 3)),
+MULTIPOLYGON Z (((0 0 0, 1 1 1, 2 3 1, 0 0 0)))
+)`,
+				`GEOMETRYCOLLECTION(
+POINT Z (2 3 4),
+LINESTRING Z (0 -1 1, 7 -1 -9),
+POLYGON Z ((0 0 7, 1 -1 -50, 2 0 0, 0 0 7)),
+MULTIPOINT Z ((2 3 1), EMPTY),
+MULTILINESTRING Z (EMPTY, EMPTY, (1 1 1, 2 2 2, 3 3 3)),
+MULTIPOLYGON Z (((0 0 0, 1 1 1, 2 3 1, 0 0 0)))
+)`,
+			},
+			expected: geom.NewGeometryCollection().MustPush(
+				geom.NewPointFlat(geom.XYZ, []float64{2, 3, 4}),
+				geom.NewLineStringFlat(geom.XYZ, []float64{0, -1, 1, 7, -1, -9}),
+				geom.NewPolygonFlat(geom.XYZ, []float64{0, 0, 7, 1, -1, -50, 2, 0, 0, 0, 0, 7}, []int{12}),
+				geom.NewMultiPointFlat(geom.XYZ, []float64{2, 3, 1}, geom.NewMultiPointFlatOptionWithEnds([]int{3, 3})),
+				geom.NewMultiLineStringFlat(geom.XYZ, []float64{1, 1, 1, 2, 2, 2, 3, 3, 3}, []int{0, 0, 9}),
+				geom.NewMultiPolygonFlat(geom.XYZ, []float64{0, 0, 0, 1, 1, 1, 2, 3, 1, 0, 0, 0}, [][]int{{12}}),
+			),
+		},
+		{
+			desc: "parse 4D geometrycollection",
+			equivInputs: []string{`GEOMETRYCOLLECTION ZM (
+POINT ZM (0 5 -10 15),
+LINESTRING ZM (0 0 0 0, 1 1 1 1),
+POLYGON ZM ((0 0 12 7, 1 -1 12 -50, 2 0 12 0, 0 0 12 7)),
+MULTIPOINT ZM ((2 -8 17 45), (0 0 0 0)),
+MULTILINESTRING ZM ((0 0 0 0, 1 1 1 1), (-2 -3 -4 -5, 0.5 -0.75 1 -1.25, 0 1 5 7)),
+MULTIPOLYGON ZM (((0 0 0 0, 1 1 1 -1, 2 3 1 -2, 0 0 0 0)))
+)`,
+				`GEOMETRYCOLLECTION(
+POINT(0 5 -10 15),
+LINESTRING(0 0 0 0, 1 1 1 1),
+POLYGON((0 0 12 7, 1 -1 12 -50, 2 0 12 0, 0 0 12 7)),
```

Comment:
can you add a test where one of the objects is `ZM` when the root is not, and vice versa?